src/rowbinary/de.rs

```diff
@@ -242,42 +400,204 @@ impl<'data> Deserializer<'data> for &mut RowBinaryDeserializer<'_, 'data> {
             }
         }
 
-        visitor.visit_seq(Access {
-            deserializer: self,
+        let inner_data_type_validator = self.validator.validate(
+            &SerdeType::Tuple,
+            &[
+                DataTypeHint::Tuple,
+                DataTypeHint::Array,
+                DataTypeHint::FixedString,
+                // FIXME: uncomment when there is a way to implement ReverseSeqAccess
+                // DataTypeHint::IPv4,
+                DataTypeHint::IPv6,
+            ],
             len,
-        })
+        )?;
+        let mut new_self = RowBinaryDeserializer {
+            input: self.input,
+            validator: inner_data_type_validator,
+        };
+        let access = RowBinaryTupleAccess {
+            deserializer: &mut new_self,
+            len,
+        };
+        visitor.visit_seq(access)
     }
 
     #[inline]
     fn deserialize_option<V: Visitor<'data>>(self, visitor: V) -> Result<V::Value> {
-        ensure_size(&mut self.input, 1)?;
+        // println!("deserialize_option call");
 
+        ensure_size(&mut self.input, 1)?;
+        let inner_data_type_validator =
+            self.validator
+                .validate(&SerdeType::Option, &[DataTypeHint::Nullable], 0)?;
         match self.input.get_u8() {
-            0 => visitor.visit_some(&mut *self),
+            0 => visitor.visit_some(&mut RowBinaryDeserializer {
+                input: self.input,
+                validator: inner_data_type_validator,
+            }),
             1 => visitor.visit_none(),
             v => Err(Error::InvalidTagEncoding(v as usize)),
         }
     }
 
     #[inline]
     fn deserialize_seq<V: Visitor<'data>>(self, visitor: V) -> Result<V::Value> {
+        // println!("deserialize_seq call");
+
+        struct RowBinarySeqAccess<'de, 'cursor, 'data, Validator>
```

Comment:
currently, seq access while being the same, are implemented separately for each related method (tuple, struct, seq), cause we have slightly different implementation details and validation rules. Perhaps can be generated via a macro